R/rvar-math.R

```diff
@@ -0,0 +1,332 @@
+#' converted to a row vector and `y` to a column vector. Numerics and logicals can be multiplied
+#' by [`rvar`]s and are broadcasted across all draws of the [`rvar`] argument. Tensor multiplication
+#' is used to efficiently multiply matrices across draws, so if either `x` or `y` is an [`rvar`],
+#' `x %**% y` will be much faster than `rdo(x %*% y)`.
+#'
+#' Because [`rvar`] is an S3 class and S3 classes cannot properly override `%*%`, [`rvar`]s use
+#' `%**%` for matrix multiplication.
+#'
+#' @return An [`rvar`] representing the matrix product of `x` and `y`.
+#'
+#' @importFrom tensorA mul.tensor as.tensor
+#' @export
+`%**%` <- function(x, y) {
+  # Fast version of rdo(x %*% y)
+
+  # ensure everything is a matrix by adding dimensions as necessary to make `x`
+  # a row vector and `y` a column vector
+  ndim_x <- length(dim(x))
+  if (ndim_x == 1) {
+    dim(x) <- c(1, dim(x))
+  } else if (ndim_x != 2) {
+    stop2("First argument (`x`) is not a vector or matrix, cannot matrix-multiply")
+  }
+
+  ndim_y <- length(dim(y))
+  if (ndim_y == 1) {
+    dim(y) <- c(dim(y), 1)
+  } else if (ndim_y != 2) {
+    stop2("Second argument (`y`) is not a vector or matrix, cannot matrix-multiply")
+  }
+
+  # convert both objects into rvars if they aren't already (this will give us
+  # a 3d draws array for each variable)
+  x <- as_rvar(x)
+  y <- as_rvar(y)
+
+  # conform the draws dimension in both variables
+  c(x, y) %<-% conform_rvar_ndraws_nchains(list(x, y))
+
+  # drop the names of the dimensions (mul.tensor gets uppity if dimension names
+  # are duplicated, but we don't care about that)
+  x_tensor <- as.tensor(draws_of(x))
+  y_tensor <- as.tensor(draws_of(y))
+  names(dim(x_tensor)) <- NULL
+  names(dim(y_tensor)) <- NULL
+
+  # do a tensor multiplication equivalent of the requested matrix multiplication
+  result <- unclass(mul.tensor(x_tensor, 3, y_tensor, 2, by = 1))
+
+  # move draws dimension back to the front
+  result <- aperm(result, c(3,1,2))
+
+  # restore dimension names (as.tensor adds dummy names to dimensions)
+  names(dim(result)) <- NULL
+  result <- copy_dimnames(draws_of(x), 1:2, result, 1:2)
+  result <- copy_dimnames(draws_of(y), 3, result, 3)
+
+  new_rvar(result, .nchains = nchains(x))
+}
+
+#' Cholesky decomposition of random matrix
+#'
+#' Cholesky decomposition of an [`rvar`] containing a matrix.
+#'
+#' @param x A 2-dimensional [`rvar`].
+#' @param ... Additional parameters passed on to `chol.tensor()`
+#'
+#' @return An [`rvar`] containing the upper triangular factor of the Cholesky
+#' decomposition, i.e., the matrix \eqn{R} such that \eqn{R'R = x}.
+#'
+#' @importFrom tensorA chol.tensor as.tensor
+#' @export
+chol.rvar <- function(x, ...) {
+  # ensure x is a matrix
+  if (length(dim(x)) != 2) {
+    stop2("`x` must be a random matrix")
+  }
+
+  # must re-order draws dimension to the end, as chol.tensor expects it there
```

Comment:
From profiling, the time in this function seems to be dominated by `tensorA::chol.tensor()` at the moment. Looking at it, `chol.tensor()` seems to be implemented using an lapply so that's not too surprising (unlike `tensorA::mul.tensor()` which I used to implement matrix multiplication, which calls down to a C implementation).

So I'd say the bigger problem is `chol.tensor()`... possibly we need to find a better tensor library and/or implement it ourselves if we want them to be performant? It looks like other functions that might make sense to have (`solve`, `svd`, ...?) will have similar issues.

If you don't object, I'd say we start by getting the initial implementation in on this PR, then can replace them with faster implementations later? Likely someone with more knowledge than I have of tensor libraries in R would be the right person to do that.

Reply:
I agree to your suggestion.